File: pcapsolution/packet-buffer-solution.go
package main

import (
  "fmt"
  "io/ioutil"
  "encoding/binary"
  "sort"
  "os"
)

/**
  There are several sections of commented out code in this source file, especially
  in the specific header parsing functions.

  Commented out code is illustrative of parsing the rest of the header values.
  The students' task does not require any of this information, so the unused
  values have been commented out. I've left the code in so that it can be shown to
  students for illustrative purposes.
**/

// These lengths are assumptions based on the specifics of the provided net.cap file.
// A real parser would need to lookup at least the value for the link layer header
// after reading the pcap file global header.
const pcapFileHeaderLen = 24
const pcapPacketHeaderLen = 16
const ethernetHeaderLen = 14

/* These two type defs make some of our code a little more explicit */

// SequenceNumber is a TCP seq or ack number, a 32 bit unsigned int
type SequenceNumber uint32

// SequenceNumberArray is an array of sequence numbers
type SequenceNumberArray []SequenceNumber

// These are required for length and sorting
func (a SequenceNumberArray) Len() int           { return len(a) }
func (a SequenceNumberArray) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a SequenceNumberArray) Less(i, j int) bool { return a[i] < a[j] }


func main() {
  // Open the file, load all the bytes into memory.
  // In some cases it would be better to stream these bytes into memory using os.Open
  pcapBytes, err := ioutil.ReadFile("net.cap")
  if err != nil {
      panic(err)
  }

  // This is another assumption from the provided net.cap -- the HTTP data was being sent to this IP
  destinationOfInterest := []byte{192, 168, 0, 101}
  payloadMap := extractFromPcap(pcapBytes, int(binary.LittleEndian.Uint32(destinationOfInterest)))

  // Sort the sequence numbers
  sequenceNumbers := make([]SequenceNumber, len(payloadMap))
  sqnIdx := 0
  for sequenceNumber := range payloadMap {
    sequenceNumbers[sqnIdx] = sequenceNumber
    sqnIdx++
  }

  // Use our custom sorter
  sort.Sort(SequenceNumberArray(sequenceNumbers))

  // get the orderd bytes in one slice
  var orderedBytes []byte
  for _, sequenceNumber := range sequenceNumbers {
    orderedBytes = append(orderedBytes, payloadMap[sequenceNumber]...)
  }

  // Find the image section within the bytes
  i := 0
  for ; i < len(orderedBytes) - 4; i++ {
    if string(orderedBytes[i:i+4]) == "\r\n\r\n" {
      i += 4 // i needs to be advanced to the first location for data
      break
    }
  }

  // Write to disk
  f, err := os.Create("./theMan.jpg")
  defer f.Close()
  if err != nil {
    panic(err)
  }

  f.Write(orderedBytes[i:])
}

// To ease in building the parser I added a simple assert function
func assertEqual(a interface{}, b interface{}, message string) {
	if a != b {
		panic(fmt.Sprintf("%s: %v != %v", message, a, b))
	}
}

/*
  Provideda filepath to a pcap file and an ip address (as a 32 bit little endian integer)
  parse the pcap and return the TCP payload for that ip address as a
  map[sequenceNumber]payloadBytes
*/
func extractFromPcap(pcapBytes []byte, destIPAddress int) map[SequenceNumber][]byte {
  packets := make(map[SequenceNumber][]byte) // Maps sequence numbers to HTTP Payload data

  globalHeader := pcapBytes[0:pcapFileHeaderLen]
  verifyPcapHeader(globalHeader)

  // Note we have to update the packetStartLocation after reading it b/c packets are variable length
  for packetStartLocation := pcapFileHeaderLen; packetStartLocation <= len(pcapBytes); {
    pcapPacketHeader := pcapBytes[packetStartLocation:packetStartLocation + pcapPacketHeaderLen]
    packetLen := parsePcapHeader(pcapPacketHeader)

    if packetLen == 0 {
      return packets
    }

    // A slice of bytes
    etherStart := packetStartLocation + pcapPacketHeaderLen
    etherEnd := etherStart + packetLen
    ethernetPacket := pcapBytes[etherStart:etherEnd]
    ipDatagram := parseEtherFrame(ethernetPacket)
    ipDestination, tcpSegment := parseIPDatagram(ipDatagram)
    sequenceNumber, tcpPayload := parseTCPMessage(tcpSegment)

    ipDestInt := int(binary.LittleEndian.Uint32(ipDestination))
    if ipDestInt == destIPAddress {
      packets[sequenceNumber] = tcpPayload
    }

    packetStartLocation += (packetLen + pcapPacketHeaderLen)
  }

  return packets
}

/*
  Given a slice representing the global PCAP header file bytes, test against the known values for our
  particular capture. If something looks wrong, the assert function panics.
*/
func verifyPcapHeader(headerSlice []byte) {
  // TODO: this typecasting seems like it could be simpler...
  assertEqual(binary.LittleEndian.Uint32(headerSlice[0:4]), uint32(0xa1b2c3d4), "Pcap Endianness")
  assertEqual(binary.LittleEndian.Uint16(headerSlice[4:6]), uint16(2), "Pcap Major Version should be 2")
  assertEqual(binary.LittleEndian.Uint16(headerSlice[6:8]), uint16(4), "Pcap Minor Version should be 4")
  assertEqual(binary.LittleEndian.Uint32(headerSlice[8:12]), uint32(0), "TZ Offset should be zero")
  assertEqual(binary.LittleEndian.Uint32(headerSlice[12:16]), uint32(0), "Timestamp Accuracy should be zero")
  assertEqual(binary.LittleEndian.Uint32(headerSlice[20:24]), uint32(1), "Link Layer Header should be 1 (for ethernet)")
}

/*
  Given a slice represneting the bytes for a Pcap Packet Header (not a global header) check that the
  total length was captured (for our test file all packets were completely captured) and return the
  length of this packet
*/
func parsePcapHeader(headerSlice []byte) int {
  totalLength := binary.LittleEndian.Uint32(headerSlice[8:12])
  capturedLength := binary.LittleEndian.Uint32(headerSlice[12:16])
  assertEqual(totalLength, capturedLength, "Pcap Packet captured length didn't match total length -- misalignment may have occured")
  return int(totalLength)
}


func parseEtherFrame(etherFrameSlice []byte) []byte {
  headerBytes := etherFrameSlice[0:ethernetHeaderLen]
  etherType := headerBytes[12:14]
  assertEqual(int(binary.LittleEndian.Uint16(etherType)), 8, "Wrong ethernet type detected") // Ether IPv4

  // destMacAddress := headerBytes[0:6]
  // sourceMacAddress := headerBytes[6:12]

  return etherFrameSlice[ethernetHeaderLen:]
}


func parseIPDatagram(datagramBytes []byte) ([]byte, []byte) {
  ipVersion := datagramBytes[0] >> 4
  assertEqual(ipVersion, byte(4), "Wrong IP version parsed")

  ipHeaderLength := 4 * (datagramBytes[0] & 7) // Masking w/ 00001111, len provided in 32bit words
  ipDest := datagramBytes[16:20]
  // ipSource := datagramBytes[12:16]

  return ipDest, datagramBytes[ipHeaderLength:]
}


func parseTCPMessage(tcpBytes []byte) (SequenceNumber, []byte) {
  seqNumber := binary.BigEndian.Uint32(tcpBytes[4:8])
  reservedSection := tcpBytes[12] & 14        // Reserved is the 3 bits within the bottom half of a single byte
  assertEqual(int(reservedSection), 0, "Reserved section is not 0, misalignment is likely")
  tcpHeaderLength := 4 * (tcpBytes[12] >> 4)  // header-size is the top half of a byte, in 32bit words

  // sourcePort := binary.BigEndian.Uint16(tcpBytes[0:2])
  // destPort := binary.BigEndian.Uint16(tcpBytes[2:4])
  // ackNumber := binary.BigEndian.Uint32(tcpBytes[8:12])
  // flagBitfield := binary.BigEndian.Uint16(tcpBytes[12:14]) & 511
  // windowSize := binary.BigEndian.Uint16(tcpBytes[14:16])
  // tcpChecksum := binary.BigEndian.Uint16(tcpBytes[16:18])
  // Ignoring urgent pointer, not actually checking checksum

  // Extract from bitfields
  // flags := map[string]bool {
  //   "ns_flag" : (flagBitfield & 256) != 0,
  //   "cwr_flag" : (flagBitfield & 128) != 0,
  //   "ece_flag" : (flagBitfield & 64) != 0,
  //   "urg_flag" : (flagBitfield & 32) != 0,
  //   "ack_flag" : (flagBitfield & 16) != 0,
  //   "psh_flag" : (flagBitfield & 8) != 0,
  //   "rst_flag" : (flagBitfield & 4) != 0,
  //   "syn_flag" : (flagBitfield & 2) != 0,
  //   "fin_flag" : (flagBitfield & 1) != 0 }


  return SequenceNumber(seqNumber), tcpBytes[tcpHeaderLength:]
}

File: pcapsolution/packet-buffer-solution.py
import os
import sys

PCAP_FILE_HEADER_LEN = 24
PCAP_PACKET_HEADER_LEN = 16
ETHERNET_HEADER_LEN = 14


def main():
    filepath = 'net.cap'
    host_of_interest = (192, 168, 0, 101)

    if len(sys.argv) == 3:
        filepath = sys.argv[1]
        host_of_interest = tuple(int(p) for p in sys.argv[2].split('.'))

    print(filepath, host_of_interest, sys.argv)
    extract_from_dump(filepath, host_of_interest)


def extract_from_dump(binary_file_path, dest_ip_tuple):
    data = {}
    with open(binary_file_path, 'rb') as f:
        global_header = f.read(PCAP_FILE_HEADER_LEN)
        verify_header(global_header)

        # The rest of the bytes will be Ethernet Frames
        while True:
            packet_header = f.read(PCAP_PACKET_HEADER_LEN)
            if not packet_header:
                break;

            packet_len = parse_pcap_header(packet_header)
            print("Handling packet, len: {}".format(packet_len))

            packet_bytes = f.read(packet_len)
            ip_datagram = parse_ether_frame(packet_bytes)
            ip_source, ip_dest, tcp_segment = parse_ip_datagram(ip_datagram)
            seq_number, tcp_payload = parse_tcp_message(tcp_segment)

            if tuple(ip_dest) == dest_ip_tuple and len(tcp_payload) > 0:
                data[seq_number] = tcp_payload

    ordered_data = [d for _, d in sorted(data.items())]
    binary_string = b''.join(ordered_data)
    relevant_data = binary_string.split(b'\r\n\r\n', 1)[1] # Everything before the split is HTTP header data

    with open('result.jpg', 'wb') as output:
        output.write(relevant_data)
    print('done')


def btoi(bs):
    return int.from_bytes(bs, 'little')


def verify_header(header_bytes):
    assert header_bytes[0:4] == bytes.fromhex('d4c3b2a1')
    assert btoi(header_bytes[4:6]) == 2
    assert btoi(header_bytes[6:8]) == 4
    assert btoi(header_bytes[8:12]) == 0
    assert btoi(header_bytes[12:16]) == 0
    print('Snapshot max size: {}'.format(btoi(header_bytes[16:20])))
    print('Link Layer Header Type: {}'.format(btoi(header_bytes[20:24])))


def parse_pcap_header(header_bytes):
    # We just make sure we got the whole packet
    total_len = btoi(header_bytes[8:12])
    captured_len = btoi(header_bytes[12:16])
    assert total_len == captured_len
    return captured_len


def parse_ether_frame(ether_frame_bytes):
    header_bytes = ether_frame_bytes[0:ETHERNET_HEADER_LEN]
    dest_mac = header_bytes[0:6]
    source_mac = header_bytes[6:12]
    ether_type = header_bytes[12:14]
    print( bytes.hex(source_mac), bytes.hex(dest_mac))
    assert ether_type == bytes.fromhex('0800') # Ether IPv4
    return ether_frame_bytes[ETHERNET_HEADER_LEN:]


def parse_ip_datagram(datagram_bytes):
     ip_version = datagram_bytes[0] >> 4
     assert ip_version == 4
     ip_header_len = 4 * (datagram_bytes[0] & 7) # Masking w/ 00001111, len provided in 32bit words

     ip_source = datagram_bytes[12:16]
     ip_dest = datagram_bytes[16:20]

     print("  src:  ", tuple(ip_source))
     print("  dest: ", tuple(ip_dest))
     print("  ip header len: ", ip_header_len)

     return ip_source, ip_dest, datagram_bytes[ip_header_len:]


def parse_tcp_message(tcp_bytes):
    source_port = big_btoi(tcp_bytes[0:2])
    dest_port = big_btoi(tcp_bytes[2:4])
    seq_number = big_btoi(tcp_bytes[4:8])
    ack_number = big_btoi(tcp_bytes[8:12])
    reserved_section = tcp_bytes[12] & 14      # Reserved is the 3 bits within the bottom half of a single byte
    tcp_header_len = 4 * (tcp_bytes[12] >> 4)  # header-size is the top half of a byte, in 32bit words
    flag_bitfield = big_btoi(tcp_bytes[12:14]) & 511
    window_size = big_btoi(tcp_bytes[14:16])
    tcp_checksum = big_btoi(tcp_bytes[16:18])
    # Ignoring urgent pointer

    # Extract from bitfields
    flags = {
        'ns_flag' : (flag_bitfield & 256) != 0,
        'cwr_flag' : (flag_bitfield & 128) != 0,
        'ece_flag' : (flag_bitfield & 64) != 0,
        'urg_flag' : (flag_bitfield & 32) != 0,
        'ack_flag' : (flag_bitfield & 16) != 0,
        'psh_flag' : (flag_bitfield & 8) != 0,
        'rst_flag' : (flag_bitfield & 4) != 0,
        'syn_flag' : (flag_bitfield & 2) != 0,
        'fin_flag' : (flag_bitfield & 1) != 0
    }

    assert reserved_section == 0

    print("  source port: ", source_port)
    print("  dest port: ", dest_port)
    print("  seq number: ", seq_number)
    print("  ack num: ", ack_number)
    print("  tcp header len: ", tcp_header_len)
    print("  window size: ", window_size)
    print("  checksum: ", tcp_checksum)

    for name, val in flags.items():
        if val:
            print("  {} set".format(name))

    return seq_number, tcp_bytes[tcp_header_len:]


def big_btoi(bs):
    return int.from_bytes(bs, 'big')

if __name__ == '__main__':
    main()
